lib/common/JSONImporter/NodeStatePatch.ts

```diff
@@ -0,0 +1,519 @@
+import {NodeSelections} from './NodeSelectors';
+import {NodeChangeSet} from './NodeChangeSet';
+import {NodeSearchUtils, setNested} from './Utils';
+import {Maybe, Result} from 'ts-monads';
+import diff, {ChangeSet, ChangeType} from 'changeset';
+import JSONImporter from "../JSONImporter";
+import Core = GmeClasses.Core;
+
+type PatchFunction = (node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections) => Promise<Result<PatchResult, PatchError>>;
+type PatchResultPromise = Promise<Result<PatchResult, PatchError>>;
+
+class PatchResult {
+    patches: NodeChangeSet | NodeChangeSet[];
+
+    constructor(patches: NodeChangeSet | NodeChangeSet[]) {
+        this.patches = patches;
+    }
+}
+
+export class PatchError extends Error {
+    constructor(msg: string) {
+        super(msg);
+    }
+}
+
+export interface PatchOperation {
+    core: GmeClasses.Core;
+    put: PatchFunction;
+    delete: PatchFunction;
+}
+
+export abstract class NodeStatePatch implements PatchOperation {
+    core: GmeClasses.Core;
+    nodeSearchUtils: NodeSearchUtils
+
+    constructor(core: GmeClasses.Core, nodeSearchUtils: NodeSearchUtils) {
+        this.core = core;
+        this.nodeSearchUtils = nodeSearchUtils;
+    }
+
+    keyLengthValidator(change) {
+        return change.key.length === 2;
+    }
+
+    async put(node: Core.Node, change: NodeChangeSet, resolvedSelectors: NodeSelections): PatchResultPromise {
+        return this._validChange(change)
```

Comment:
This is an example of what I meant. Each patch operation just defines their own `_put` and `_delete` methods which accept a validated change.  Validation can be defined in another method, `_validChange`. (I am sure there is a better name - maybe `getValidChange` or something?)

Reply:
There are two changes I had to make:

1. There should be two validators: `validChangePut` and `validChangeDelete`. 
2. Not every error can be resolved (deletion for MemberAttributes for example with the above signature).